### PR TITLE
Include leaderelection in client-go; 

### LINKED ISSUE #39117: Missing "leaderelection" package when staging to client-go
"pkg/client/leaderelection" depends on a couple other client packages. But it doesn't sit in client-go repo right now.

It's not possible to use "leaderelection" package when using purely "client-go". The staging/copy.sh should also copy "leaderelection" packages.

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #39117
Fix https://github.com/kubernetes/client-go/issues/28

This PR:
* includes the leaderelection to the staging client-go
* to avoid conflict with golang's testing package, renames package /testing to /testutil, and renames cache/testing to cache/testframework

```release-note
client-go now includes the leaderelection package
```

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@k8s-bot bazel test this

**Comment by caesarxuchao** [comment]
@mikedanese do you know why the bazel build is failing? The "Details" doesn't provide any information.

**Comment by caesarxuchao** [comment]
@k8s-bot bazel test this

**Comment by deads2k** [comment]
leader election currently requires access to `endpoints` which is too much power to grant generally.  We shouldn't move leader election until we have a resource specifically for it which we can grant fine-grained access to.

**Comment by caesarxuchao** [comment]
```
cmd/kube-controller-manager/app/options/options.go:100: cannot use leaderelection.DefaultLeaderElectionConfiguration() (type "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/componentconfig".LeaderElectionConfiguration) as type "k8s.io/kubernetes/pkg/apis/componentconfig".LeaderElectionConfiguration in field value
cmd/kube-controller-manager/app/options/options.go:214: cannot use &s.KubeControllerManagerConfiguration.LeaderElection (type *"k8s.io/kubernetes/pkg/apis/componentconfig".LeaderElectionConfiguration) as type *"k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/componentconfig".LeaderElectionConfiguration in argument to leaderelection.BindFlags
# k8s.io/kubernetes/federation/cmd/federation-controller-manager/app/options
federation/cmd/federation-controller-manager/app/options/options.go:101: cannot use leaderelection.DefaultLeaderElectionConfiguration() (type "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/componentconfig".LeaderElectionConfiguration) as type "k8s.io/kubernetes/pkg/apis/componentconfig".LeaderElectionConfiguration in field value
federation/cmd/federation-controller-manager/app/options/options.go:132: cannot use &s.ControllerManagerConfiguration.LeaderElection (type *"k8s.io/kubernetes/pkg/apis/componentconfig".LeaderElectionConfiguration) as type *"k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/componentconfig".LeaderElectionConfiguration in argument to leaderelection.BindFlags
# k8s.io/kubernetes/plugin/cmd/kube-scheduler/app/options
plugin/cmd/kube-scheduler/app/options/options.go:85: cannot use &s.KubeSchedulerConfiguration.LeaderElection (type *"k8s.io/kubernetes/pkg/apis/componentconfig".LeaderElectionConfiguration) as type *"k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/componentconfig".LeaderElectionConfiguration in argument to leaderelection.BindFlags
```

We hit these issues because client-go is copying the types (see https://github.com/kubernetes/kubernetes/issues/44065). I'm actively solving that issue.

**Comment by pmorie** [comment]
@caesarxuchao any progress on the client-go issue?

**Comment by caesarxuchao** [comment]
WIP: https://github.com/kubernetes/kubernetes/pull/44784

**Comment by caesarxuchao** [comment]
@pmorie @hongchaodeng @deads2k could you take a look? Thanks.

File: pkg/client/leaderelectionconfig/config.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+)
+
+const (
+	DefaultLeaseDuration = 15 * time.Second
+	DefaultRenewDeadline = 10 * time.Second
+	DefaultRetryPeriod   = 2 * time.Second
+)
+
+func DefaultLeaderElectionConfiguration() componentconfig.LeaderElectionConfiguration {
```
Comment (caesarxuchao) [inline pkg/client/leaderelectionconfig/config.go]: I left the componentconfig part in kubernetes. Other parts are moved to client-go.

**Comment by caesarxuchao** [comment]
Tests passed. @pmorie @hongchaodeng @deads2k PTAL. Thanks.

**Comment by hongchaodeng** [comment]
/lgtm

Thanks @caesarxuchao for the work.

**Pull request merged**